include/aws/common/byte_buf.h

```diff
@@ -35,5 +36,49 @@ static inline struct aws_byte_buf aws_byte_buf_from_array(const uint8_t c_str[],
     return (struct aws_byte_buf){.buffer = (uint8_t *)&c_str[0], .len = len};
 }
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * No copies, no string allocations. Fills in output with a list of aws_byte_buf instances where buffer is
+ * an offset into the input_str and len is the length of that string in the original buffer.
+ *
+ * Edge case rules are as follows:
+ * if the string begins with split_on, an empty string will be the first entry in output
+ * if the input string has two adjacent split_on tokens, an empty string will be inserted into the output.
+ * if the input string ends with split_on, it will be ignored.
+ *
+ * It is the user's responsibility to properly initialize output. Recommended number of preallocated elements from output
```

Comment:
What happens if output is fixed-sized and too small to accommodate all of the tokens?

Also, it's very useful to have a split with a limited number of tokens, where any additional tokens after the first N are left appended to the last token.

Reply:
AWS_OP_ERR will be returned and the error will be in aws_last_error().

I'm not sure I see the usecase for your second <p/>

Reply:
Example usecase: I have a S3 path and I want to split it into bucket and key. I have a CSV line and I really only care about the third field. Etc.